pkg/controller/container-runtime-config/helpers.go

```diff
@@ -539,24 +539,93 @@ func validateRegistriesConfScopes(insecure, blocked, allowed []string, icspRules
+	itmsRuleSourcePolicy := map[string]apicfgv1.MirrorSourcePolicy{}
+	for _, idms := range idmsRules {
+		for _, mirrorSet := range idms.Spec.ImageDigestMirrors {
+			// keep the field (source, mirror) empty checking for ImageContenSourcePolicy objects
+			// ImageDigestMirrorSet has schema validation
 			if mirrorSet.Source == "" {
 				return fmt.Errorf("invalid empty entry for source configuration")
 			}
-			if strings.Contains(mirrorSet.Source, "*") {
-				return fmt.Errorf("wildcard entries are not supported with mirror configuration %q", mirrorSet.Source)
+			if mirrorSet.MirrorSourcePolicy != "" {
+				if policy, ok := idmsRuleSourcePolicy[mirrorSet.Source]; ok {
+					if policy != mirrorSet.MirrorSourcePolicy {
+						return fmt.Errorf("conflicting mirrorSourcePolicy is set for the same source %q in imagedigestmirrorsets", mirrorSet.Source)
+					}
+				} else {
+					idmsRuleSourcePolicy[mirrorSet.Source] = mirrorSet.MirrorSourcePolicy
+				}
 			}
+
 			for _, mirror := range mirrorSet.Mirrors {
 				if mirror == "" {
 					return fmt.Errorf("invalid empty entry for mirror configuration")
 				}
-				if strings.Contains(mirror, "*") {
+				if strings.Contains(string(mirror), "*") {
 					return fmt.Errorf("wildcard entries are not supported with mirror configuration %q", mirror)
 				}
+				if mirrorSet.MirrorSourcePolicy == apicfgv1.NeverContactSource && string(mirror) == mirrorSet.Source {
+					return fmt.Errorf("cannot set mirrorSourcePolicy: NeverContactSource if the source %q is one of the mirrors", mirrorSet.Source)
+				}
 			}
 		}
+	}
 
+	for _, itms := range itmsRules {
+		for _, mirrorSet := range itms.Spec.ImageTagMirrors {
+			if mirrorSet.MirrorSourcePolicy != "" {
+				if policy, ok := itmsRuleSourcePolicy[mirrorSet.Source]; ok {
+					if policy != mirrorSet.MirrorSourcePolicy {
+						return fmt.Errorf("conflicting mirrorSourcePolicy is set for the same source %q in imagetagmirrorsets", mirrorSet.Source)
+					}
+				} else {
+					itmsRuleSourcePolicy[mirrorSet.Source] = mirrorSet.MirrorSourcePolicy
+				}
+				if policy, ok := idmsRuleSourcePolicy[mirrorSet.Source]; ok {
+					if policy != mirrorSet.MirrorSourcePolicy {
+						return fmt.Errorf("conflicting mirrorSourcePolicy is set for the same source %q in imagedigestmirrorsets and imagetagmirrorsets", mirrorSet.Source)
+					}
+				}
+
+			}
+			for _, mirror := range mirrorSet.Mirrors {
+				if mirrorSet.MirrorSourcePolicy == apicfgv1.NeverContactSource && mirror == apicfgv1.ImageMirror(mirrorSet.Source) {
+					return fmt.Errorf("cannot set mirrorSourcePolicy: NeverContactSource if the source %q is one of the mirrors", mirrorSet.Source)
+				}
+			}
+		}
 	}
 	return nil
 }
+
+// mergeICSPtoIDMS converts the spec of ImageContentSourcePolicy to the spec of ImageDigestMirrorSet
+// and appends the converted icspRules to given idmsRules
+func mergeICSPtoIDMS(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy,
```

Comment:
See https://github.com/openshift/runtime-utils/pull/15#discussion_r842734297 — I don’t think every single mirror consumer (and there are quite a few) should copy&paste this.

Reply:
Yes, I agree... Moved this helper to runtime-utils.